### PR TITLE
Restrict access to Bank's HardForks

### PR DESCRIPTION
**Pull request by steviez**
#### Problem
Callers could previously obtain a a lock to read/write HardForks from any Bank. This would allow any caller to modify, and creates the opportunity for inconsistent handling of what is considered a valid hard fork (ie too old).

#### Summary of Changes
This PR adds a function to Bank so consistent sanity checks can be applied; the caller will already have a Bank as that is where they would have obtained the HardForks from in the first place. Additionally, change the getter to return a copy of HardForks (simple Vec).

### CODE REVIEW

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -7308,8 +7307,24 @@ impl Bank {
...
-    pub fn hard_forks(&self) -> Arc<RwLock<HardForks>> {
-        self.hard_forks.clone()
+    /// Get a snapshot of the current set of hard forks
+    pub fn hard_forks(&self) -> HardForks {
+        self.hard_forks.read().unwrap().clone()
+    }
+
+    pub fn register_hard_fork(&self, new_hard_fork_slot: Slot) {
+        let bank_slot = self.slot();
+        if new_hard_fork_slot > bank_slot {
```
Comment (mvines) [inline runtime/src/bank.rs:7317]: It's actually ok to use the current slot as long as the bank isn't frozen

use case: first restart attempt fails, and so we try again at the same slot by adding a second `--hard-fork` argument. Although looking at at the original code, I'm doubting myself that we actually currently do the right thing when `--hard-fork <SLOTX> --hard-fork <SLOTX>` is specified! That 100% worked in the past though as we have had multiple testnet restart attempt on the same slot before.

```rust
let _lock = self.freeze_lock()
if new_hard_fork_slot > bank_slot || (new_hard_fork_slot == bank_slot && !self.is_frozen()) {
  ..
}

File: core/src/validator.rs [hunk 2]
```diff
@@ -665,19 +665,15 @@ impl Validator {
             entry_notifier,
             Some(poh_timing_point_sender.clone()),
         )?;
+        let hard_forks = bank_forks.read().unwrap().root_bank().hard_forks();
```
Comment (mvines) [inline core/src/validator.rs:668]: Did you intentionally change this from `working_bank()` to `root_bank()`?

File: runtime/src/bank.rs (hunk 1 above)
Comment (steviez) [reply runtime/src/bank.rs:7317]: > It's actually ok to use the current slot as long as the bank isn't frozen

Good point.

> I'm doubting myself that we actually currently do the right thing when --hard-fork <SLOTX> --hard-fork <SLOTX> is specified! That 100% worked in the past though as we have had multiple testnet restart attempt on the same slot before.

The Nov 2022 testnet outage had a handful of retry attempt at the same slot (this was the Hetzner-related one where we used your idea to warp slot to get dead nodes out of leader rotation & turbine tree after several failed attempts). When making subsequent snapshots for restart attempts, we added an extra `--hard-fork` at the same slot.

I think the key was that each restart snapshot was being created from the last snapshot of "normal" validator operation; this made it such that the hard forks were registered with an older bank and ready when it came time to freeze the restart slot bank in `Bank::hash_internal_state()`:
https://github.com/solana-labs/solana/blob/875b95a07bc398a3dc6f158f7e8eaac7e721311f/runtime/src/bank.rs#L7581-L7586

Comment (steviez) [reply runtime/src/bank.rs]: Just pushed a commit to allow hard fork at bank slot if the bank is not frozen.

As a note, I deviated slightly from your suggestion and did not use `is_frozen()`. `is_frozen()` would attempt to momentarily grab a second read lock while the initial one is held. As of Rust 1.62, `RwLock` gives preference to writers which means another thread grabbing the write lock (via `freeze()`) after `freeze_lock()` but before `is_frozen()` would deadlock.

Comment (mvines) [reply runtime/src/bank.rs]: > As of Rust 1.62, RwLock gives preference to writers which means another thread grabbing the write lock (via freeze()) after freeze_lock() but before is_frozen() would deadlock.

🤔 but `let _lock = self.freeze_lock();` in my code snippet would still hold a read lock when `is_frozen()` attempts to acquire that second read lock, so no writer could get in.

File: core/src/validator.rs (hunk 2 above)
Comment (mvines) [reply core/src/validator.rs:668]: wfm, thanks for the explanation

File: runtime/src/bank.rs (hunk 1 above)
Comment (steviez) [reply runtime/src/bank.rs]: > which means another thread grabbing the write lock (via freeze()) after freeze_lock() but before is_frozen() would deadlock.

Re-reading my comment, this might have been more clear if I would have said "another thread **attempting to grab** the write lock ...". Let me try again; here is a diagram of the scenario I'm describing with lettered threads and numbered "events":
```rust
// Thread A                             // Thread B    
// 1 - Thread 1 obtains a read lock
bank.register_hard_fork()
    let _lock = self.freeze_lock();

                                        // 2 - Thread 2 attempts to obtain write lock in freeze
                                        bank.freeze();
                                            let mut hash = self.hash.write().unwrap();

// 3 - Thread 1 attempts to
//     obtain a read lock again
    ... self.is_frozen() ...
```
- Thread A obtains a read lock (1) just fine
- Thread B attempts to obtain a write lock (2), but can't due to (1) so it waits
- Thread A attempts to recursively obtain read lock (3)
    - In Rust < 1.62 on Linux, the second read lock attempt would be given priority over the write lock attempt and proceed immediately. The writer could then go after the initial read lock (1) is eventually released (assuming no other readers swoop in)
    - In Rust >= 1.62 on Linux, `RwLock` changes to prefer writers to avoid starvation. So, the attempt to grab second read lock (3) may get blocked waiting for the write lock attempt (2) to complete. But that write lock attempt is blocked by the already held read lock (1).

The docs mention this ([see Potential deadlock example](https://doc.rust-lang.org/std/sync/struct.RwLock.html)) with a diagram that I modeled mine after. [Here's a tweet](https://twitter.com/m_ou_se/status/1526211117651050497?lang=en) from one of the Rust engineers on the matter too.

Comment (mvines) [reply runtime/src/bank.rs]: Got it! Thanks, that's sneaky.

**Review by mvines — APPROVED** [review]
